src/main/java/io/phdata/streamliner/schemadefiner/attributeretriever/AttributeRetrieverFactory.java

```diff
@@ -0,0 +1,42 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+package io.phdata.streamliner.schemadefiner.attributeretriever;
+
+import java.sql.Connection;
+import java.util.function.Supplier;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class AttributeRetrieverFactory {
+
+  private static final Logger log = LoggerFactory.getLogger(AttributeRetrieverFactory.class);
+
+  public static AttributeRetriever getAttributeRetriever(
+      String dbType, final Supplier<Connection> connectionSupplier) {
+    switch (dbType) {
+      case "hive":
+        return new HiveAttributeRetriever(connectionSupplier.get());
+      case "hive2":
+        return new HiveAttributeRetriever(connectionSupplier.get());
+      case "impala":
+        return new HiveAttributeRetriever(connectionSupplier.get());
+      default:
+        log.info("Attribute retriever not available for {}", dbType);
+        return null;
```

Comment:
Returning `null` here is a hint that there is a design problem. Fetching Hive tables and Snowflake tables is different, so they should have different classes. This `null` is a dependency, and becomes application logic, which is hard to reason about.

This refactor might not be worth making now. Can this logic be pushed closer to where it's used? There's this factory returning a class (or null), that gets pushed into a class where it's used in a method. Can we just make this a method in that class instead? It might be the issue that you don't have the connectionSupplier available?

Reply:
Agree. I should not return null in the default case. Rather I should create a class like `NoAttributeRetriever` and return this.
If we make this as a method in a different class, it will get limited to that class only. I think as a factory class, this can be used anywhere in the application.